Custom syntax highlighting theme has been removed / Spyder crashes when a custom syntax highlighting theme has as color an invalid hex value

Hi

I noticed that my personal (dark) theme was recently removed (maybe after an update) ; twice, i tried to re-create my theme:

1. new values don't change anything
2. when spyder is closed and reopened, the window immediatly closes when it appears
3. when using the anaconda prompt to launch spyder, no message is displayed

I needed to reset and return to default settings to use Spyder

Paul

* Spyder version: 6.0.2  (conda)
* Python version: 3.11.10 64-bit
* Qt version: 5.15.2
* PyQt5 version: 5.15.10
* Operating System: Windows-10-10.0.19045-SP0

Thank you for the new info @paul18frr ! Adding a `!` to one of the colors of an scheme, I was able to reproduce crashing Spyder after it is relaunched. Checking, from a dev version I was able to get the following traceback:

```python-traceback
Traceback (most recent call last):
  File "E:\Acer\Documentos\Spyder\spyder\bootstrap.py", line 214, in <module>
    start.main()
  File "E:\Acer\Documentos\Spyder\spyder\spyder\app\start.py", line 285, in main
    mainwindow.main(options, args)
  File "E:\Acer\Documentos\Spyder\spyder\spyder\app\mainwindow.py", line 1425, in main
    mainwindow = create_window(MainWindow, app, splash, options, args)
  File "E:\Acer\Documentos\Spyder\spyder\spyder\app\utils.py", line 370, in create_window
    main.post_visible_setup()
  File "E:\Acer\Documentos\Spyder\spyder\spyder\app\mainwindow.py", line 861, in post_visible_setup
    plugin.on_mainwindow_visible()
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\plugin.py", line 562, in on_mainwindow_visible
    self.create_new_client(give_focus=False)
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\plugin.py", line 687, in create_new_client
    self.get_widget().create_new_client(
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\widgets\main_widget.py", line 1696, in create_new_client
    client = ClientWidget(
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\widgets\client.py", line 148, in __init__
    self.shellwidget = ShellWidget(
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\widgets\shell.py", line 167, in __init__
    super().__init__(*args, **kw)
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\widgets\debugging.py", line 206, in __init__
    super(DebuggingWidget, self).__init__(*args, **kwargs)
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\widgets\debugging.py", line 85, in __init__
    super(DebuggingHistoryWidget, self).__init__(*args, **kwargs)
  File "e:\acer\documentos\spyder\spyder\external-deps\qtconsole\qtconsole\rich_jupyter_widget.py", line 71, in __init__
    super().__init__(*args, **kw)
  File "e:\acer\documentos\spyder\spyder\external-deps\qtconsole\qtconsole\jupyter_widget.py", line 129, in __init__
    self._syntax_style_changed()
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\widgets\shell.py", line 1439, in _syntax_style_changed
    self._highlighter._style = create_style_class(self.syntax_style)
  File "E:\Acer\Documentos\Spyder\spyder\spyder\plugins\ipythonconsole\utils\style.py", line 171, in create_style_class
    class StyleClass(Style):
  File "C:\Users\dalth\anaconda3\envs\spyder-dev\lib\site-packages\pygments\style.py", line 122, in __new__
    ndef[0] = colorformat(styledef)
  File "C:\Users\dalth\anaconda3\envs\spyder-dev\lib\site-packages\pygments\style.py", line 79, in colorformat
    assert False, f"wrong color format {text!r}"
AssertionError: wrong color format '#b0e686!'
```

--- a/spyder/plugins/appearance/widgets.py
+++ b/spyder/plugins/appearance/widgets.py
@@ -72,23 +72,25 @@ class SchemeEditor(QDialog):
 
     def validate_colors(self):
         """
-        Validate the current color scheme and display a message box listing 
+        Validate the current color scheme and display a message box listing
         any invalid colors.
         """
         invalid_colors = {}
         scheme_name = self.last_used_scheme
-        pattern = r"^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3}|[A-Fa-f0-9]{8}|[A-Fa-f0-9]{4})$"
+        pattern = (
+            r"^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3}|[A-Fa-f0-9]{8}|[A-Fa-f0-9]{4})$"
+        )
         for key in self.widgets[scheme_name]:
             items = self.widgets[scheme_name][key]
 
             if not bool(re.match(pattern, items[0].text())):
-                    invalid_colors[key] = items[0].text()
+                invalid_colors[key] = items[0].text()
 
         if invalid_colors:
             message = _("The following properties have invalid colors:\n\n")
             for property_name, color in invalid_colors.items():
                 name = syntaxhighlighters.COLOR_SCHEME_KEYS[property_name]
-                clean_name = name[:-1].replace("<br>","")
+                clean_name = name[:-1].replace("<br>", "")
                 message += _(
                     "The property <b>{}</b> has an invalid color: {}\n"
                 ).format(clean_name, color)
@@ -100,7 +102,7 @@ class SchemeEditor(QDialog):
                 QMessageBox.Ok,
                 self
             )
-            msgbox.exec_()         
+            msgbox.exec_()
         else:
             self.accept()
 

--- a/spyder/plugins/appearance/widgets.py
+++ b/spyder/plugins/appearance/widgets.py
@@ -77,11 +77,13 @@ class SchemeEditor(QDialog):
                     invalid_colors[key] = items[0].text()
 
         if invalid_colors:
-            message = _("The following properties have invalid colors:\n")
+            message = _("The following properties have invalid colors:\n\n")
             for property_name, color in invalid_colors.items():
+                name = syntaxhighlighters.COLOR_SCHEME_KEYS[property_name]
+                clean_name = name[:-1].replace("<br>","")
                 message += _(
                     "The property {} has an invalid color: {}.\n"
-                    ).format(property_name.upper(), color)
+                    ).format(clean_name, color)
             title = _('Error getting colors.')
             self.msgbox = QMessageBox(
                         QMessageBox.Warning,

--- a/spyder/plugins/appearance/widgets.py
+++ b/spyder/plugins/appearance/widgets.py
@@ -8,7 +8,10 @@
 
 from qtpy.QtCore import Qt
 from qtpy.QtWidgets import (QDialog, QDialogButtonBox, QGridLayout, QGroupBox,
-                            QHBoxLayout, QVBoxLayout, QWidget)
+                            QHBoxLayout, QVBoxLayout, QWidget, QMessageBox)
+from qtpy import QtWidgets
+
+import re
 
 from spyder.api.translations import _
 from spyder.api.widgets.dialogs import SpyderDialogButtonBox
@@ -42,7 +45,8 @@ class SchemeEditor(QDialog):
         self.setLayout(layout)
 
         # Signals
-        bbox.accepted.connect(self.accept)
+        bbox.accepted.connect(self.validate_colors)
+        #bbox.accepted.connect(self.accept)
         bbox.accepted.connect(self.get_edited_color_scheme)
         bbox.rejected.connect(self.reject)
 
@@ -59,6 +63,39 @@ class SchemeEditor(QDialog):
         scheme creation.
         """
         return self.scheme_name_textbox[self.last_used_scheme].text()
+
+    def validate_colors(self):
+        """
+        Get the values of the last edited color scheme to be used in an instant
+        preview in the preview editor, without using `apply`.
+        """
+        invalid_colors = {}
+        scheme_name = self.last_used_scheme
+        pattern = r"^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3}|[A-Fa-f0-9]{8}|[A-Fa-f0-9]{4})$"
+        for key in self.widgets[scheme_name]:
+            items = self.widgets[scheme_name][key]
+
+            if(not bool(re.match(pattern, items[0].text()))):
+                    invalid_colors[key] = items[0].text()
+
+        if invalid_colors:
+            message = _("The following properties have invalid colors:\n")
+            for property_name, color in invalid_colors.items():
+                message += _(
+                    "The property {} has an invalid color: {}.\n"
+                    ).format(property_name.upper(), color)
+            title = _('Error getting colors.')
+            self.msgbox = QMessageBox(
+                        QMessageBox.Warning,
+                        title,
+                        message,
+                        QMessageBox.Ok,
+                        self
+                    )
+            self.msgbox.exec_()         
+        else:
+            self.accept()
+
 
     def get_edited_color_scheme(self):
         """

--- a/spyder/plugins/appearance/widgets.py
+++ b/spyder/plugins/appearance/widgets.py
@@ -6,11 +6,19 @@
 
 """Appearance entry in Preferences."""
 
-from qtpy.QtCore import Qt
-from qtpy.QtWidgets import (QDialog, QDialogButtonBox, QGridLayout, QGroupBox,
-                            QHBoxLayout, QVBoxLayout, QWidget, QMessageBox)
-
 import re
+
+from qtpy.QtCore import Qt
+from qtpy.QtWidgets import (
+    QDialog,
+    QDialogButtonBox,
+    QGridLayout,
+    QGroupBox,
+    QHBoxLayout,
+    QMessageBox,
+    QVBoxLayout,
+    QWidget,
+)
 
 from spyder.api.translations import _
 from spyder.api.widgets.dialogs import SpyderDialogButtonBox
@@ -73,7 +81,7 @@ class SchemeEditor(QDialog):
         for key in self.widgets[scheme_name]:
             items = self.widgets[scheme_name][key]
 
-            if(not bool(re.match(pattern, items[0].text()))):
+            if not bool(re.match(pattern, items[0].text())):
                     invalid_colors[key] = items[0].text()
 
         if invalid_colors:
@@ -82,17 +90,17 @@ class SchemeEditor(QDialog):
                 name = syntaxhighlighters.COLOR_SCHEME_KEYS[property_name]
                 clean_name = name[:-1].replace("<br>","")
                 message += _(
-                    "The property {} has an invalid color: {}.\n"
-                    ).format(clean_name, color)
-            title = _('Error getting colors.')
-            self.msgbox = QMessageBox(
-                        QMessageBox.Warning,
-                        title,
-                        message,
-                        QMessageBox.Ok,
-                        self
-                    )
-            self.msgbox.exec_()         
+                    "The property <b>{}</b> has an invalid color: {}\n"
+                ).format(clean_name, color)
+
+            msgbox = QMessageBox(
+                QMessageBox.Warning,
+                _('Error setting colors'),
+                message,
+                QMessageBox.Ok,
+                self
+            )
+            msgbox.exec_()         
         else:
             self.accept()
 

--- a/spyder/plugins/appearance/widgets.py
+++ b/spyder/plugins/appearance/widgets.py
@@ -9,7 +9,6 @@
 from qtpy.QtCore import Qt
 from qtpy.QtWidgets import (QDialog, QDialogButtonBox, QGridLayout, QGroupBox,
                             QHBoxLayout, QVBoxLayout, QWidget, QMessageBox)
-from qtpy import QtWidgets
 
 import re
 
@@ -46,7 +45,6 @@ class SchemeEditor(QDialog):
 
         # Signals
         bbox.accepted.connect(self.validate_colors)
-        #bbox.accepted.connect(self.accept)
         bbox.accepted.connect(self.get_edited_color_scheme)
         bbox.rejected.connect(self.reject)
 
@@ -66,8 +64,8 @@ class SchemeEditor(QDialog):
 
     def validate_colors(self):
         """
-        Get the values of the last edited color scheme to be used in an instant
-        preview in the preview editor, without using `apply`.
+        Validate the current color scheme and display a message box listing 
+        any invalid colors.
         """
         invalid_colors = {}
         scheme_name = self.last_used_scheme

